README.md
netrc
=====

Simple netrc files manager

[![Build Status](https://travis-ci.org/fduch/netrc.svg?branch=master)](https://travis-ci.org/fduch/netrc)

Installation
--------------
To install netrc manager into your project please use composer:

```sh
php composer.phar require fduch/netrc
```

Usage
-----
Parsing system-wide or custom netrc is quite simple:

``` php
<?php

use Fduch\Netrc\Netrc;
use Fduch\Netrc\Exception\ParseException;

try {
    // you can specify path to netrc file as an argument of Netrc::parse() method
    $parsed = Netrc::parse();
    // dumps key-value array corresponding to machine.one entry
    var_dump($parsed['machine.one']);
} catch (ParseException $e) {
    //something is wrong with your netrc file
}

```


CHANGELOG.md
CHANGELOG
=========

### 1.0.0 (2015-06-12)

  * First major release


tests/bootstrap.php
<?php
/**
 * This file is part of the Netrc package.
 *
 * (c) Alex Medvedev <alex.medwedew@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Date: 3/21/14
 */

$loader = require __DIR__ . "/../vendor/autoload.php";
$loader->addPsr4('Fduch\\Netrc\\', __DIR__.'/');

tests/ParserTest.php
<?php
/**
 * This file is part of the Netrc package.
 *
 * (c) Alex Medvedev <alex.medwedew@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Date: 3/21/14
 */

namespace Fduch\Netrc;

use PHPUnit_Framework_TestCase;

class ParserTest extends \PHPUnit_Framework_TestCase
{
    protected $parser;

    protected function setUp()
    {
        $this->parser = new Parser();
    }

    /**
     * @test
     */
    public function correctNetrcParsedSuccessfully()
    {
        $netrc = <<<EOF
machine machine.one login john password pass1
machine machine.two login steve password pass2
EOF;
        $this->assertEquals(
            $this->parser->parse($netrc),
            array(
                "machine.one" => array('login' => 'john', 'password' => 'pass1'),
                "machine.two" => array('login' => 'steve', 'password' => 'pass2'),
            )
        );
    }

    /**
     * @test
     */
    public function emptyLinesAndWhiteSpacesParsedSuccessfully()
    {
        $netrc = <<<EOF


machine          machine.one login john password         pass1
machine


 machine.two
login steve
    password pass2


EOF;
        $this->assertEquals(
            $this->parser->parse($netrc),
            array(
                "machine.one" => array('login' => 'john', 'password' => 'pass1'),
                "machine.two" => array('login' => 'steve', 'password' => 'pass2'),
            )
        );
    }

    /**
     * @expectedException Fduch\Netrc\Exception\ParseException
     * @test
     */
    public function incorrectDefaultThrowsException()
    {
        $netrc = <<<EOF
default machine.one login john password pass1
EOF;
        $this->parser->parse($netrc);
    }

    /**
     * @expectedException Fduch\Netrc\Exception\ParseException
     * @test
     */
    public function incorrectKeyThrowsException()
    {
        $netrc = <<<EOF
machine machine.one logOn john password pass1
EOF;
        $this->parser->parse($netrc);
    }

    /**
     * @expectedException Fduch\Netrc\Exception\ParseException
     * @test
     */
    public function emptyMachineNameThrowsException()
    {
        $netrc = <<<EOF
machine
EOF;
        $this->parser->parse($netrc);
    }

    /**
     * @expectedException Fduch\Netrc\Exception\ParseException
     * @test
     */
    public function unsetValueForValidKeyThrowsException()
    {
        $netrc = <<<EOF
machine machine.one login
EOF;
        $this->parser->parse($netrc);
    }

    /**
     * @expectedException Fduch\Netrc\Exception\ParseException
     * @test
     */
    public function attemptToSetInfoForUnknownMachineThrowsException()
    {
        $netrc = <<<EOF
login john
EOF;
        $this->parser->parse($netrc);
    }

    /**
     * @test
     */
    public function commentsInNetrcAreIgnored()
    {
        $netrc = <<<EOF
machine machine.one login john password pass1

# machine machine.two
#login machine.two
machine machine.two login steve password pass2 #should be omited

#machine machine.two login steve password pass2
machine machine.three login mike
     password pass3#should be omited
# machine machine.two
#login machine.two
# steve machine.two


EOF;
        $this->assertEquals(
            $this->parser->parse($netrc),
            array(
                "machine.one"   => array('login' => 'john', 'password' => 'pass1'),
                "machine.two"   => array('login' => 'steve', 'password' => 'pass2'),
                "machine.three" => array('login' => 'mike', 'password' => 'pass3'),
            )
        );
    }
}
 

src/Netrc.php
<?php
/**
 * This file is part of the Netrc package.
 *
 * (c) Alex Medvedev <alex.medwedew@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Date: 3/21/14
 */

namespace Fduch\Netrc;

use Fduch\Netrc\Exception\ParseException;

/**
 * Netrc manager
 *
 * @author Alex Medvedev
 */
class Netrc
{
    /**
     * Parses netrc file specified or default one
     *
     * @param string|null $filename
     *
     * @throws ParseException when netrc file could not be read or parsed
     *
     * @return array of netrc values grouped by machines
     */
    public static function parse($filename = null)
    {
        // fetch netrc filename if it is not specified
        if (!$filename) {
            if (!$homePath = getenv('HOME')) {
                throw new ParseException("HOME environment variable must be set for correctly netrc handling");
            }
            if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
                $filename = strtr($homePath, '\\', '/') . '/_netrc';
            } else {
                $filename = rtrim($homePath, '/') . '/.netrc';
            }
        }
        $filename = realpath($filename);

        // check that netrc file is available
        if (!file_exists($filename) || !is_readable($filename) || !$content = file_get_contents($filename)) {
            throw new ParseException("netrc file does not exist or is not readable");
        }

        // parse file
        $parser = new Parser();
        return $parser->parse($content);
    }
}
 

src/Parser.php
<?php
/**
 * This file is part of the Netrc package.
 *
 * (c) Alex Medvedev <alex.medwedew@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * Date: 3/21/14
 */

namespace Fduch\Netrc;

use Fduch\Netrc\Exception\ParseException;

/**
 * Parses netrc file
 *
 * @author Alex Medvedev
 */
class Parser
{
    /**
     * List of netrc keys
     *
     * @var array
     */
    protected static $netrcKeys = array("machine", "default", "login", "password", "account", "macdef");

    /**
     * Parses netrc content
     *
     * @param string $content file content
     *
     * @throws ParseException when netrc file could not be parsed
     *
     * @return array of netrc values grouped by machines
     */
    public function parse($content)
    {
        $tokens = $this->getTokens($content);

        $i = 0;
        $result = array();
        $currentMachine = '';

        while ($i < count($tokens)) {
            $currentToken = $tokens[$i];
            if ($currentToken == 'machine') {
                if (!isset($tokens[$i+1])) {
                    throw new ParseException("Cannot fetch machine name");
                }
                $currentMachine = $tokens[$i+1];
                $result[$currentMachine] = array();
                $i+=2;
            } else if ($currentToken == 'default') {
                $currentMachine = 'default';
                $result[$currentMachine] = array();
                $i++;
            } else {
                if (!in_array($currentToken, static::$netrcKeys)) {
                    throw new ParseException("Cannot handle unknown key '$currentToken' for '$currentMachine' machine");
                }
                if (!isset($tokens[$i+1])) {
                    throw new ParseException("Cannot fetch value of '$tokens[$i]' key for '$currentMachine' machine");
                }
                if (!$currentMachine) {
                    throw new ParseException("Cannot handle key '$currentToken' for unset machine");
                }
                $result[$currentMachine][$currentToken] = $tokens[$i+1];
                $i+=2;
            }
        }

        return $result;
    }

    /**
     * Fetches significative (not inside inline comments) tokens from the original content
     *
     * @param $content
     *
     * @return array of tokens
     */
    private function getTokens($content)
    {
        // fetch non-empty lines
        $lines = preg_split('/\r\n|\r|\n/', $content, -1, PREG_SPLIT_NO_EMPTY);
        $tokens = array();
        foreach ($lines as $line) {
            // throwing tokens after '#' sign
            $commentCharPosition = strpos($line, "#");
            $lineWithoutComments = ($commentCharPosition === false) ? $line : substr($line, 0, strpos($line, "#"));
            // storing current line tokens
            if ($lineWithoutComments) {
                $tokens = array_merge($tokens, preg_split('/\s+/', $lineWithoutComments, -1, PREG_SPLIT_NO_EMPTY));
            }
        }
        return $tokens;
    }
}
 